README.md
# Edith 监控平台script

上报前端错误到Edith后台

[更新日志](./CHANGELOG.md)

---

## 开发

[开发请看](./DEVELOPMENT.md)

## 安装

```bash
npm install @inke-design/edith-script -save
```

或者CDN引用，脚本一直会更新到最新版本，不需要更换链接，也支持放在其他域名路径下，[下载cdn源码](https://webcdn.inke.cn/edith.cn/edith.cn.zip)

建议放在`<head>`里，以保证错误和记录数据更加准确

```javascript
<script src="https://webcdn.inke.cn/edith.cn/edith.min.js"></script>
```

## 使用方式

- npm方式

```javascript
// entry.js

import Edith from '@inke-design/edith-script'

Edith.init({
  apiKey: 'apikey', // 用于区分不同项目
  silentDev: true, // 开发环境下不上报，根据域名是否为ip或者localhost来判断
  plugins: [ // 内置插件
    'breadcrumbs', // 记录用户行为堆栈
    'redo', // 记录录屏
    'network', // 记录网络信息
  ]
})
```

- CDN方式

```javascript
window.Edith.init({
  apiKey: 'apikey', // 用于区分不同项目
  silentDev: true, // 开发环境下不上报，根据域名是否为ip或者localhost来判断
  plugins: [ // 内置插件
    'breadcrumbs', // 记录用户行为堆栈
    'redo', // 记录录屏
    'network', // 记录网络信息
  ]
})
```

## 其他参数

```javascript
Edith.init({
  apiKey: 'apikey', // 用于区分不同项目，根据域名是否为ip或者localhost来判断
  silentDev: true, // 开发环境下不上报
  plugins: [ // 内置插件，选择引入
    'breadcrumbs', // 记录用户行为堆栈
    'redo', // 记录录屏
    'network', // 记录网络信息
  ],
  resourceWhiteList: ['www.baidu.com', /^((?!(a\.com|b\.com)).)*$/], // 资源加载监听的白名单（支持正则，字符串会忽略判断协议和query参数）
  ajaxWhiteList: ['//example.com/search'], // 网络请求监听的白名单（支持正则，字符串会忽略判断协议和query参数）
  silentPromise: true, // 是否不监听Promise错误，默认为false
  silentResource: true, // 是否不监听资源加载错误，默认为false
  silentHttp: true, // 是否需要不监控网络请求异常，默认为false
  silentWebsocket: true, // 是否不监听Websocket错误，默认为false
  setHttpBody: true, // 是否上报http请求里的body，默认为false
  filters: err => { // 过滤错误函数，参数为带有name, title, url, message, ajax, target的错误信息字段, 返回值为真值，则不上报错误（自定义上报不拦截）
    if(err.message === 'Script error.') return true
  }
})
```

## 自定义上报方法

在某些情况下，因一些无法跟踪的业务错误，导致业务出现问题，可以采用自定义上报该错误。
Edith内部有一个`debug`方法，可以传两个参数，第一个参数为错误名称(字符串)，第二个参数为错误信息，可以是字符串或者对象等。

  ```javascript
  if(uid === undefined) Edith.debug('NoUID', {
    msg: 'uid为空',
    ...
  })

  ```

## 录屏隐私处理

有些业务，可能由于涉及到隐私和不宜公开的信息，不希望录屏上报到后台，需要给指定的`DOM`做隐私处理。

- 在`HTML`元素中添加类名`.edith-block`将会避免该元素及其子元素被录制，回放时取而代之的是一个同等宽高的占位元素。
- 在`HTML`元素中添加类名`.edith-ignore`将会避免录制该元素的输入事件。
- `input[type="password"]`类型的密码输入框默认不会录制输入事件。

## filters参数示例

1. 过滤name为 TypeError 的错误

  ```javascript
  Edith.filters = err => {
    if(err.name === 'TypeError') return true
  }

  ```

2. 过滤 method 为 GET，且 status 为 403 的错误

  ```javascript
  Edith.filters = err => {
    if(err.ajax.method === 'GET' && err.ajax.stutus === 403) return true
  }
  ```

3. 过滤指定域名的资源加载错误

  ```javascript
  Edith.filters = err => {
    return err.name === 'resourceError' && err.message.match(/example.com/)
  }
  ```

4. 过滤url为空字符串的图片加载错误

  ```javascript
  Edith.filters = err => {
    const { target: { tagName = '' }, message } = err
    return tagName.match(/img/g) && message === ''
  }
  ```


CHANGELOG.md
# 2021-06-21 v1.0.0

<!-- - 更新脚本接入方式（自定义事件监听方式），防止脚本加载影响页面加载 -->
- 优化同一错误短时间内反复提交的问题，增加定时批量上报
- 增加`performance`插件，可查看页面的性能各项指标
- 增加获取本apiKey的白名单
<!-- - 用重写`onpopstate`来实现url变化 -->

# 2021-06-07 v0.4.6

- 修复`img`的src为空报错时，message错误的问题

# 2021-06-05 v0.4.5

- 修复获取`responseText`失败的问题

# 2021-06-03 v0.4.0~v0.4.4

- 优化数据上报量，压缩数据，优化原子参数过长造成的数据冗余的问题
- 优化录屏带来的内存占用问题
- 优化插件包体积大小
- 增加脚本自身错误上报
- 优化message类型
- 优化部分已知错误

# 2021-05-25 v0.3.3&v0.3.4

- 压缩上报数据量,plugins, redo, breadcrumbs都进行压缩
- 解决自定义事件的timeStamp不准的问题
- 不在打包cdn分版本号

# 2021-05-19 v0.3.2

- 修复自定义上报传对象导致message错误的问题
- 修复filters参数报错，没有catch住的问题

# 2021-05-13 v0.3.1

- 支持更好的自测
- 支持在其他域名路径下引用本脚本

# 2021-02-18 v0.3.0

- 进一步缩小录屏插件的上报数据
- 修复`unhandledrejection`错误无法上报的问题

# 2020-12-16 v0.2.8

- 修复用户行为的fetch请求没有记录到url的问题
- 修复用户行为的记录网络请求数据有误的情况
- 增加对外链js报错的解决方案，给外链js增加`crossorigin='anonymous'`的属性
- 修复`WebSocket`记录数据不准的问题

# 2020-12-9 v0.2.7(v0.2.6 ignore)

- 修复用户行为堆栈数据过大的问题

# 2020-12-4 v0.2.5

- 优化script标签的属性性能，适配预渲染
- 缩小录屏插件的上报数据
- 优化自定义插件数据的限制
- 增加上报脚本版本号
- 解决img标签为空时的报错信息
- 优化行为堆栈白名单的判断，减少数据量
- 异常数据中不包含性能数据
- 增加过滤sockjs-node的常见调试报错

# 2020-11-18 0.2.4（v0.2.3有问题）

- 优化Edith对象的属性及其原型链展示
- 优化行为堆栈的bug
- 加快监控错误的时机

# 2020-09-21 v0.2.2

- 修复对Promise的polyfill的判断

# 2020-09-16 v0.2.1

- 增加silentHttp是否监听网络错误的参数

# 2020-09-16 v0.2.0

- 自定义插件数据，字段统一处理
- 增加上报cookie

# 2020-09-16 v0.1.11

- 修复http请求header和body的获取
- 增加webSokect监听错误和堆栈记录
- 增加错误过滤，以及控制Promise监听，资源加载监听、webSocket、是否上报http的body的参数
- 修复Promise不支持的解决方法
- 修复开发环境下不能上报自定义错误的bug

# 2020-08-22 v0.1.10

- 修复网络请求报错的判断
- 修复部分bug

# 2020-8-19 v0.1.6

- 修复插件安装使用问题

# 2020-8-19 v0.1.2

- 修复插件打包bug

# 2020-8-19 v0.1.1

- 修复自检bug

# 2020-8-19 v0.0.9

- 上报错误后，state需要清空，防止产生无用数据

# 2020-8-19 v0.0.8

- 修复npm包的问题

# 2020-8-19 v0.0.7

- 优化生命周期，增加自检
- 解决请求记录没有时间戳的问题

# 2020-8-18 v0.0.6

- 重构代码逻辑，拓展功能改为插件引入

# 2020-8-11 v0.0.5

- 更改请求地址
- 请求错误监听的逻辑

# 2020-7-30 v0.0.4

- 更改检测网速的请求方法
- 优化部分包大小
- 增加主动埋点上报功能


src/utils.js
// /* eslint-disable no-undef */
import { pack } from './pack'
import md5 from './md5'
import { reportScriptError, batchAdd } from './api'
import { cdnUrl } from './config'
import { removeHttpAndQuery, getCurrentTime } from './common'
const performance = window.performance
/**
 * 参数格式化, 符合url方式
 * @params {Object} {a: '123', age: '18'}
 * @return {String} 'a=123&age=18'
 */
export const stringifyParams = (params, cb) => {
  let name
  let value
  let str = ''

  for (name in params) {
    value = params[name]
    str += name + '=' + (
      typeof cb === 'function' ?
        cb(value, name) :
        value && typeof value === 'object' && !Object.getPrototypeOf(value).slice ? JSON.stringify(value) : value ) + '&'
  }
  return str.slice(0, -1)
}

// 获取ua环境信息
export const uaInfo = (() => {
  const uaInfo = navigator.userAgent.match(/[^(\s]+(\s\([^)]+\))?/g)
  return uaInfo
})()


// 对方法进行封装，防止内部报错
export const tryCatchFn = (fn, edith) => function(...args) {
    try {
      typeof fn !== 'function' && (fn = () => {})
      return fn.apply(this, args)
    } catch (error) {
      console.warn('edith 内部报错', error)
      reportEdithError('EdithError', error, edith)
    }
  }

// 通用事件监听方法
export const edithAddEventListener = (name, fn, useCapture) => {
  var ae = window.attachEvent
  if (addEventListener) {
    // 所有主流浏览器，除了 IE 8 及更早版本
    addEventListener(name, tryCatchFn(fn), useCapture)
  } else if (ae) {
    // IE 8 及更早版本
    ae(`on${name}`, tryCatchFn(fn))
  }
}


// 执行环境监测以及性能
export const getPerform = () => {
  // const getEntriesByType = 'getEntriesByType'
  if (!performance) {
    // console.warn('edith 内部报错: 当前环境不支持性能监控', )
    return false
  }
  return {
    timing: performance.timing,
    // chromeLoadingTiming: chrome && chrome.loadTimes(),
    // entriesTiming: {
    //   navigation: per[getEntriesByType]('navigation'),
    //   paint: per[getEntriesByType]('paint'),
    //   resource: per[getEntriesByType]('resource').filter(item => !item.name.match(/hm\.baidu\.com\/hm\.gif/)),
    // },
  }
}

// 异步加载scripts标签
export const loadScript = (url, cb, reject = () => {}) => {
  const script = document.createElement('script')
  script.setAttribute('cdn-rendered', '')
  script.src = url
  script.onload = cb
  script.onerror = reject
  document.getElementsByTagName('head')[0].appendChild(script)
  // document.head.appendChild(script)
}
// 异步加载插件，封装成promise
export const loadCdnScript = (url, name) =>
  new Promise((resolve, reject) => {
    loadScript(url, () => resolve({ default: window.Edith[name] }), reject)
  })

// 避免map方法进入reject
export const getPromiseResult = (promises) => {
  const handlePromise = Promise.all(
    promises.map((fn) => (fn.catch ? fn.catch((err) => 0) : fn))
  )
  return handlePromise
}

// 得到Headers对象里的数据
export const getHeaders = h => {
  const header = {}
  if (h instanceof Headers) {
    for (var p of h) {
      header[p[0]] = p[1]
    }
    return header
  }
  return h
}
const MORE_DATA_TIP = '[large size Edith Plugins Data]'
const MIN_SIZE = 6 * 1024 * 1024 // 4kb
export const minSize = tryCatchFn((params) => {
  const res = {}
  for (var name in params) {
    var value = params[name]
    res[name] = transToString(value) > MIN_SIZE
          ? MORE_DATA_TIP
          : value
  }
  return res
})
// 如果是Error，则取stack
export function transToString(p) {
  return (p && typeof p === 'object' && p instanceof Error ? p.stack : p) || ''
}
// 压缩对象成加密压缩的字符串
export function compressData (p) {
  p = pack(transToString(p))
  return p
}
let apiKey = ''
export function reportEdithError (name, err, edith) {
  edith && edith.apiKey && (apiKey = edith.apiKey)
  reportScriptError({
    version: EDITH_VERSION,
    format: FORMAT,
    name,
    message: compressData(err),
    scriptUrl: cdnUrl,
    userAgent: navigator.userAgent,
    apiKey
  })
}


/**
 * 把数据保存到本地
 */
 export const saveLocalData = tryCatchFn((key, item) => {
  localStorage.setItem(key, JSON.stringify(item))
})
/**
 * 读取本地数据
 */
export const getLocalData = tryCatchFn((key) => JSON.parse(localStorage.getItem(key)))
// 将每个错误生成唯一key，用错判断是否为重复错误
export function getOnlyTag(err) {
  const { message, name, extraInfo } = err
  const _massage = name === 'resourceError' ? removeHttpAndQuery(message || '') : message
  const tag = location.host + location.pathname.replace(/index\.html$/, '') + name
    + _massage + removeHttpAndQuery(extraInfo.url || '') + (extraInfo.method || '')
  return md5(tag)
}

const Edith_KEY = '_edith_key'
const TIMEOUT = 10 * 60 * 1000 // 十分钟
// 检测本地是否有已经上报的数据
export function checkLocalData(tag) {
  const localData = getLocalData(Edith_KEY) || {}
  const res = !!localData[tag] // 本地有记录，就不上报
  if(res) {
    localData[tag].push(getCurrentTime()) // 本地记录，不上报
  } else {
    localData[tag] = [ getCurrentTime() ]  // 上报且本地计数加1
  }
  saveLocalData(Edith_KEY, localData)
  return res
}
const EXPIRATION_TIME = 2 * 60 * 60 * 1000 // 错误报存过期时间
function task(edith) {
  const localData = getLocalData(Edith_KEY) || {}
  const now = getCurrentTime()
  const params = {
    apiKey: edith.apiKey,
    errs: [],
    now,
    cookie: document.cookie,
    userAgent: navigator.userAgent,
  }
  Object.keys(localData).forEach(tag => {
    const originLenth = localData[tag].length
    const rest = localData[tag].filter(time => now - time < EXPIRATION_TIME) // 拿到有效时间内的错误
    if (rest.length === 1 && originLenth === 1) return
    const times = rest.slice(originLenth === rest.length ? 1 : 0)
    params.errs.push({
      onlyTag: tag,
      times
    }) // 上报
    delete localData[tag]
  })
  if(params.errs.length === 0) return
  batchAdd(params)
  saveLocalData(Edith_KEY, localData)
}
// 
export function loopTask(edith, init) {
  !init && task(edith)
  let st = setTimeout(function() {
    task(edith)
    clearTimeout(st)
    st = null
    loopTask(edith, 1)
  }, TIMEOUT)
}

/*
// 汉字转unicode
export const ch2Unicode = (data) => {
  if (data === '') return ''
  const res = data.replace(/([\u4E00-\u9FA5]|[\uFE30-\uFFA0])/g, function(
    char
  ) {
    return '\\u' + char.charCodeAt(0).toString(16)
  })
  return res
}
// unicode转汉字
export const unicode2Ch = (asciicode) => {
  asciicode = asciicode.split('\\u')
  var nativeValue = asciicode[0]
  for (var i = 1; i < asciicode.length; i++) {
    var code = asciicode[i]
    nativeValue += String.fromCharCode(parseInt('0x' + code.substring(0, 4)))
    if (code.length > 4) {
      nativeValue += code.substring(4, code.length)
    }
  }
  return nativeValue.replace(/^[\t\r\v\n\f\0]+/, '')
}
*/
